profiler/profile.go

```diff
@@ -112,43 +189,59 @@ func (b *batch) addProfile(p *profile) {
 	b.profiles = append(b.profiles, p)
 }
 
-func (p *profiler) runProfile(t ProfileType) (*profile, error) {
-	switch t {
-	case HeapProfile:
-		return heapProfile(p.cfg)
-	case CPUProfile:
-		return cpuProfile(p.cfg)
-	case MutexProfile:
-		return mutexProfile(p.cfg)
-	case BlockProfile:
-		return blockProfile(p.cfg)
-	case GoroutineProfile:
-		return goroutineProfile(p.cfg)
-	case expGoroutineWaitProfile:
-		return goroutineWaitProfile(p.cfg)
-	case MetricsProfile:
-		return p.collectMetrics()
-	default:
-		return nil, errors.New("profile type not implemented")
+func (p *profiler) runProfile(t ProfileType) ([]*profile, error) {
+	var (
+		pt    = t.collector()
+		start = now()
+	)
+	data, err := pt.Collect(pt, p)
+	if err != nil {
+		return nil, err
 	}
-}
 
-// writeHeapProfile writes the heap profile; replaced in tests
-var writeHeapProfile = pprof.WriteHeapProfile
+	profs := []*profile{{
+		name: pt.Filename,
+		data: data,
+	}}
 
-func heapProfile(cfg *config) (*profile, error) {
-	var buf bytes.Buffer
-	start := now()
-	if err := writeHeapProfile(&buf); err != nil {
-		return nil, err
+	if pt.Delta {
+		p1, err := pprofile.Parse(bytes.NewReader(data))
+		if err != nil {
+			return nil, fmt.Errorf("delta prof parse: %v", err)
+		}
+
```

Comment:
I've generally tried to follow the same style as other Go repositories, reducing the amount of empty lines, specifically around curly braces (which already automatically provide enough spacing). Can we remove it?

Reply:
I'd be okay to reduce the vertical whitespace I use, but I'm wondering under which circumstances would you find vertical whitespace acceptable within a function body?

My personal style is perhaps a bit idiosyncratic, but the underlaying pattern is that I separate by logical chunks that I want to be able to easily navigate to. In vim `shift+]` and `shift+[` provides a very convenient way to jump between blocks of code separated by vertical whitespace.

Reply:
I tried to put my thoughts in point 3 here: https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines. But I am very happy to reassess, clarify and change anything as long as it makes sense.